### PR TITLE
Alexmont/filter adapter

### PR DESCRIPTION
**Pull request by alexmontAmazon**
## What does this PR do?

- added FilterAdapter to filter out DOM bits in a specified source Adapter
- refactored DPE to use an invisible root node for simplicity and maintainability
- now handling patches where PropertyHandlers change their underlying type
- fixed bug where ElidingLabel text wasn't updated because it ignores underlying QLabel text
- switched ints to size_t where logical
- removed unnecessary KeyQueryDPE ui include

## How was this PR tested?
Locally with a new CVar editor, coming soon! (tm)
FilterAdapterComplete

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 1]
```diff
@@ -715,19 +648,37 @@ namespace AzToolsFramework
...
                 // check if it's a PropertyHandler; if it is, just set it from the DOM directly
-                auto foundEntry = m_widgetToPropertyHandler.find(childWidget);
-                if (foundEntry != m_widgetToPropertyHandler.end())
+                auto foundEntry = m_widgetToPropertyHandlerInfo.find(childWidget);
+                if (foundEntry != m_widgetToPropertyHandlerInfo.end())
                 {
-                    foundEntry->second->SetValueFromDom(valueAtSubPath);
+                    auto handlerId = AZ::Interface<PropertyEditorToolsSystemInterface>::Get()->GetPropertyHandlerForNode(valueAtSubPath);
+
+                    // check if this patch has morphed the PropertyHandler into a different type
```
Comment (amzn-tmryan) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:656]: Nice. Did you run into a bug here or just notice this when refactoring?

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 2]
```diff
@@ -765,6 +716,95 @@ namespace AzToolsFramework
...
+                widgetAlignment = Qt::AlignRight;
+                break;
+            }
+            m_columnLayout->WidgetAlignment(columnWidget, widgetAlignment);
+        }
+
+        //! If the sharePrior attribute is present, add the previous widget to the column layout.
+        //! Set the SharePrior boolean so we know to create a new shared column layout, or add to an existing one
+        auto sharePrior = AZ::Dpe::Nodes::PropertyEditor::SharePriorColumn.ExtractFromDomNode(domValue);
+        if (sharePrior.has_value() && sharePrior.value())
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:753]: nit: You can use `value_or(false)` here

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 3]
```diff
@@ -765,6 +716,95 @@ namespace AzToolsFramework
...
+        }
+
+        // If the UseMinimumWidth attribute is present, add the widget to set of widgets using their minimum width
+        auto minimumWidth = AZ::Dpe::Nodes::PropertyEditor::UseMinimumWidth.ExtractFromDomNode(domValue);
+        if (minimumWidth.has_value() && minimumWidth.value())
+        {
+            m_columnLayout->AddMinimumWidthWidget(columnWidget);
+        }
+
+        // insert after the found index; even if nothing were found and priorIndex is -1,
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:770]: uber nit: even if nothing *was* found

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 4]
```diff
@@ -608,15 +531,22 @@ namespace AzToolsFramework
...
         }
     }
 
     void DPERowWidget::HandleOperationAtPath(const AZ::Dom::PatchOperation& domOperation, size_t pathIndex)
     {
         const auto& fullPath = domOperation.GetDestinationPath();
         auto pathEntry = fullPath[pathIndex];
-        AZ_Assert(pathEntry.IsIndex() || pathEntry.IsEndOfArray(), "the direct children of a row must be referenced by index");
+
+        const bool isValidIndex = pathEntry.IsIndex() || pathEntry.IsEndOfArray();
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:543]: Could you clarify when this is used? It's fully valid to patch a property of a row (e.g. to update the `ForceAutoExpand` property), but it looks like this is maybe only called with the root of a patch so you're just saying the very first patch entry must be an index because we don't have any adapter node attributes that would end up here which... I think is true, but it might be possible to make this clearer.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 1 above)
Comment (NicholasVanSickle) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:656]: +1 This is a really nice catch. For reference, this is a fully valid patch that normally would infrequently occur; changing a PropertyEditor from a Slider to a SpinBox or whatever isn't something we normally do, but the way patch generation works right now it's possible to add or remove row entries and have the generated patches be replaces instead. I have some thoughts on how to improve this going forward (by specifying a comparison key when comparing sets of nodes) but it's good to have this path working!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 5]
```diff
@@ -715,19 +648,37 @@ namespace AzToolsFramework
...
                 else
                 {
-                    QLabel* changedLabel = qobject_cast<QLabel*>(childWidget);
+                    auto changedLabel = qobject_cast<AzQtComponents::ElidingLabel*>(childWidget);
                     AZ_Assert(changedLabel, "not a label, unknown widget discovered!");
                     if (changedLabel)
                     {
                         auto labelString = AZ::Dpe::Nodes::Label::Value.ExtractFromDomNode(valueAtSubPath).value_or("");
-                        changedLabel->setText(QString::fromUtf8(labelString.data(), aznumeric_cast<int>(labelString.size())));
+                        changedLabel->SetText(QString::fromUtf8(labelString.data(), aznumeric_cast<int>(labelString.size())));
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:681]: 😞 that's a bit of an API footgun...

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+
+    bool RowFilterAdapter::IsRow(const Dom::Value& domValue)
+    {
+        return (domValue.IsNode() && domValue.GetNodeName() == Dpe::GetNodeName<Dpe::Nodes::Row>());
+
+    }
+
+    bool RowFilterAdapter::IsRow(const Dom::Path& sourcePath) const
+    {
+        auto sourceRoot = m_sourceAdapter->GetContents();
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp:53]: nit: DOM copies are quite cheap compared to the lookup but it's valid to `const auto&` these guys and avoid it entirely here

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.h [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        void UpdateMatchState(MatchInfoNode* rowState);
+
+        //! returns the first path in the ancestry of sourcePath that is of type Row, including self
+        Dom::Path GetRowPath(const Dom::Path& sourcePath) const;
+
+        //! indicates whether all children of a direct match are considered matching as well
+        bool m_includeAllMatchDescendents = true;
+        bool m_filterActive = false;
+
+        MatchInfoNode* m_root = nullptr;
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.h:92]: I don't see a dtor so I suspect this leaks memory (I only see a delete in the reset method). I'd suggest just making this an `AZStd::unique_ptr` (might also be worth doing with the child nodes too if we end up running into any funkiness there)

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+                else
+                {
+                    // removed node wasn't a row, so we need to re-cache the owning row's cached info
+                    CacheDomInfoForNode(sourceContents[rowPath], matchingRow);
+                    UpdateMatchState(matchingRow);
+                }
+            }
+            else if (operationIterator->GetType() == AZ::Dom::PatchOperation::Type::Replace)
+            {
+                // replace operations can change child values from or into rows. Handle all cases
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp:185]: There's... an argument that JSON-patch was too permissive a delta comms structure, but this is well handled.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp [hunk 9]
```diff
@@ -0,0 +1,501 @@
...
+                        auto matchingRow = GetMatchNodeAtPath(rowPath);
+                        CacheDomInfoForNode(sourceContents[rowPath], matchingRow);
+                        UpdateMatchState(matchingRow);
+                    }
+                }
+
+            }
+            else if (operationIterator->GetType() == AZ::Dom::PatchOperation::Type::Add)
+            {
+                // <apm> test this! Does an add imply shifting all subsequent nodes back by 1, or are we given those messages?
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp:223]: Add should imply pushing all nodes back by 1, I believe - I believe but am not 100% sure we generate only append-to-end add patches right now for Node children which lets you get away with this.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+                AZ_Error("FilterAdapter", false, "patch operation not supported yet");
+            }
+        }
+        NotifyResetDocument();
+    }
+
+    void RowFilterAdapter::HandleDomMessage(const AZ::DocumentPropertyEditor::AdapterMessage& message,[[maybe_unused]] Dom::Value& value)
+    {
+        // forward all messages unaltered
+        #undef SendMessage // for now, work around windows #defines. We should probably just rename our SendMessage
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp:249]: +1 to renaming the method to not cause pain - this also might need an `#ifdef` for non-Windows platforms but I'm sure AR will be happy to weigh in

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp [hunk 11]
```diff
@@ -0,0 +1,501 @@
...
+    ValueStringFilter::ValueStringFilter()
+        : RowFilterAdapter()
+    {
+    }
+
+    void ValueStringFilter::SetFilterString(QString filterString)
+    {
+        if (m_filterString != filterString)
+        {
+            m_filterString = std::move(filterString);
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp:407]: nit: `AZStd::move` (a linter for AZStd/std violations might be kind of cool?)

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp [hunk 12]
```diff
@@ -0,0 +1,501 @@
...
+        return (m_filterString.isEmpty() || actualNode->m_matchableDomTerms.contains(m_filterString, Qt::CaseInsensitive));
+    }
+
+    void ValueStringFilter::StringMatchNode::AddStringifyValue(const Dom::Value& domValue)
+    {
+        QString stringifiedValue;
+
+        if (domValue.IsNull())
+        {
+            stringifiedValue = QStringLiteral("null");
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp:468]: Not 100% sure we'll want this, but I think it's fine to leave it as I'm not sure we'll ever even have a null value field.

**Review by NicholasVanSickle — COMMENTED** [review]
Really happy to see this, a lot of DPE and Adapter functionality is coming together here!

I have a few little bits of feedback (only one blocking in the form of a likely memory leak), but I would suggest diving into some testing stories as follow-up work, as I think both the FilterAdapter and the DPE itself would benefit from unit testing, both the FilterAdapter and the DPE now have several patching corner cases that could be validated and guarded against regression with unit tests (even if they're just regression tests for a few of the scary cases to start with)

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 3 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:770]: ahh, but this is not a plurality case, the "were" is from often misused subjunctive mood! I respect the effort though.
https://www.rd.com/article/if-i-was-vs-if-i-were/

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 1 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:664]: @amzn-tmryan sure did! It turns out filtering is a really good stress test for patching code

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 5 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:689]: meh, the lowercase setText does exist on elided labels, but all it does is mask (not override) the QLabel version. I'm just calling the uppercase version here for clarity. The real gotcha was realizing that the QLabel version of setText, on an elided label, does not set the text.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 2 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp]: done, thanks

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 4 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:551]: yikes. No, this is a real bug, albeit one we've always had. I'll add some code to address this.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp (hunk 6 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp]: thanks! Fixed in multiple places now.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp (hunk 8 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp:185]: ah yes, an argument from a bygone era. We'll just have to keep going with what we've got...

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp (hunk 11 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp]: fixed, thanks!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp (hunk 12 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp:471]: fair, but it doesn't hurt. Resolving...

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 13]
```diff
@@ -686,7 +686,7 @@ namespace AzToolsFramework
                     if (changedLabel)
                     {
                         auto labelString = AZ::Dpe::Nodes::Label::Value.ExtractFromDomNode(valueAtSubPath).value_or("");
-                        changedLabel->SetText(QString::fromUtf8(labelString.data(), aznumeric_cast<int>(labelString.size())));
+                        changedLabel->setText(QString::fromUtf8(labelString.data(), aznumeric_cast<int>(labelString.size())));
```
Comment (amzn-tmryan) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:689]: Just making sure you meant to change this back to `setText` after seeing your comment.

**Review by amzn-tmryan — APPROVED** [review]
Approved pending that outstanding comments are handled as you see fit, and confirmation that the call to "SetText" is supposed to be "setText" again.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.h (hunk 7 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.h:93]: I tried going the unique_ptr route, but since there's a deque of them to manage the children it became really unwieldy, and maybe impossible. I think the destructor covers it. Good catch!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp (hunk 10 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp]: renamed to SendAdapterMessage and undef removed

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 13 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp]: both are good.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 4 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp]: fixed, now clearing row and setting it from the corresponding DOM node. Might be overkill, but since this happens seldom to never, I think safety is the right approach.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp (hunk 9 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp]: cool, fixed this by adding a replace parameter to PopulateNodesAtPath since the caller should always know if it's an add or replace.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 3 above)
Comment (NicholasVanSickle) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:783]: I'll concede that you're technically grammatically correct and disagree and commit on this being idiomatic and helpful usage of subjunctive mood 😜

**Review by NicholasVanSickle — APPROVED** [review]
